File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        HashMap<Character, Integer> map = new HashMap<>();
        List<Integer> result = new ArrayList<>();

        for(int i = 0; i < p.length(); i++){
            char ch = p.charAt(i);

            map.put(ch, map.getOrDefault(ch, 0) + 1);
        }
        int match = 0;
        for(int i = 0; i < s.length(); i++){
            char incoming = s.charAt(i);

            if(map.containsKey(incoming)){
                int cnt = map.get(incoming);
                cnt -= 1;
                map.put(incoming, cnt);
                if(cnt == 0) match++;
            }

            if(i >= p.length()){
                char outgoing = s.charAt(i - p.length());
                if(map.containsKey(outgoing)){
                    int cnt = map.get(outgoing);
                    cnt += 1;
                    map.put(outgoing, cnt);
                    if(cnt == 1) match--;
                }
            }

            if(match == map.size()){
                result.add(i - p.length() + 1);
            }
        }

        return result;
    }
}

// Sliding Window
// Time Complexity: O(n)
// Space Complexity: O(1)

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        hashMap = dict()
        result = []
        for char in p:
            hashMap[char] = hashMap.get(char, 0) + 1
        
        match = 0
        for i in range(len(s)):
            char = s[i]

            if char in hashMap.keys():
                cnt = hashMap.get(char)
                cnt -= 1
                hashMap[char] = cnt
                if(cnt == 0):
                    match += 1
            
            if i >= len(p):
                if s[i - len(p)] in hashMap.keys():
                    cnt = hashMap.get(s[i - len(p)])
                    cnt += 1
                    hashMap[s[i - len(p)]] = cnt
                    if cnt == 1:
                        match -= 1
            
            if match == len(hashMap):
                result.append(i - len(p) + 1)
        
        return result